waspls/src/Wasp/Backend/ConcreteParser/Internal.hs

```diff
@@ -0,0 +1,585 @@
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE LambdaCase #-}
+
+module Wasp.Backend.ConcreteParser.Internal
+  ( -- * Internal parsing library
+    GrammarRule,
+    parse,
+
+    -- * Combinators
+    (-->),
+    as,
+    group,
+    perhaps,
+    -- | `(<>)` is used to add choice to a parser.
+    --
+    -- __Example:__
+    --
+    -- >>> let number = int <> double
+    -- >>> parse [T.Token T.Int 2] number
+    -- (...,[],[SyntaxNode (Token T.Int) 2 []])
+    -- >>> parse [T.Token  T.Double 3] number
+    -- (...,[],[SyntaxNode (Token T.Double) 2 []])
+    (<>),
+
+    -- * Primitive grammar rules
+
+    -- | These grammar rules (besides `eof`) consume a single token of the kind
+    -- matching their name. The syntax node they produce has an "S.Token" kind,
+    -- unless otherwise specified.
+    lparen,
+    rparen,
+    lsquare,
+    rsquare,
+    lcurly,
+    rcurly,
+    comma,
+    colon,
+    kwImport,
+    kwFrom,
+    kwTrue,
+    kwFalse,
+    string,
+    int,
+    double,
+    lquote,
+    rquote,
+    quoted,
+    -- | A primitive for "T.Identifier" is not included: this should always be
+    -- done with an `as`, i.e. @"T.Identifier" ``as`` "S.DictKey"@
+    eof,
+
+    -- * Future improvements
+
+    -- | TODO: First "easy" place for more improvement is to add the same type
+    -- of error recovery in the "Group" grammar rule to all other primitive grammar rule:
+    --
+    -- > Check if token we error on can be matched by any following grammar rule, if
+    -- > so, don't consume that token. Otherwise, do consume it.
+    --
+    -- While this works well in many cases, it can do what feels like the
+    -- wrong thing in some cases. For example, consider the following invalid
+    -- wasp code:
```

Comment:
So is this the second improvement then?

It is just that it is not clear in this block of comments what is this referring to.
How should I put it -> I read the first TODO now, it says we should move error recovery from group to chain, and then it starts with the quoted part "Check if token we error on ..." -> why is that quoted? Is that a separate TODO? Is it still part of the same TODO (there is : at the end of first TODO so that indicates it continues?)? Is this "Check if token..." -> is this a new TODO, or is it explaining how stuff works right now? Ah, ok actually I get it now, it explains how things work right now!
Ok let's do it differently, let's do smth like this;
```hs
    -- |
    -- TODO #1:
    -- ========
    -- First improvement is to move error recovery from the group rule
    -- to the '-->' rules.
    --
    --
    -- TODO #2:
    -- ========
    -- Right now, when doing error recovery, we check if token we errored on
    -- can be matched by any following grammar rule, if so, we leave the token
    -- for the following rule. Otherwise, discard the token.
    -- 
    -- While this works well in many cases, it can do what feels like the
    -- wrong thing in some cases. For example, consider the following invalid
    -- wasp code:
    --
```

Reply:
Hm now I realized I am bothering you with this while you are probably just planning to move it into GH issues once you are close to being done.